### PR TITLE
feat: Delete intermediary accounts

### LINKED ISSUE #1797: Log Message - ERR no delegation distribution info
## Background

During the network restart using ``v10.0.0`` our logs displayed only the following message immediately after restarting ``osmosisd``.

```ERR no delegation distribution info```

Upon [reporting this](https://discord.com/channels/798583171548840026/842550346865180682/985578355648774174) to the Osmosis Validators discord channel, @ValarDragon [explained ](https://discord.com/channels/798583171548840026/842550346865180682/985578461596880947)the message was due to "no swaps in that pool" and asked me to create this issue so the log messaging could be improved.

## Suggested Design

TBD

## Acceptance Criteria

TBD

### PR DESCRIPTION
**Pull request by mattverse**
Closes #1797 

## What is the purpose of the change

We currently do not delete intermediary accounts without any delegations, instead keeping them all in store. 
The problem is that we even iterate them during epoch, which is the main cause of the log `ERR no delegation distribution info`, as we're trying to get rewards from delegations that doesn't exist.

This PR introduces logic to deleting intermediary accounts, where we delete an intermediary account when there's no intermediary account connection and delegations from the intermediary account.

Upon my scope of investigation, I do not find any points that can be troublesome when deleting an intermediary account, although there were potential concerns for deleting intermediary account. (Note that we're not deleting the account itself using accountKeeper, we're only deleting this from the superfluid module state)

This PR achieves the goal of deleting unused intermediary accounts by adding the deleting logic and check inside EndBlocker of the superfluid module. (This has to be done in superfluid module not the lockup module because of go lang circular dependency issue. We can get an get around with this via interfaces but tbd upon reveiws)

## Brief Changelog

- Add logic for checking and deleting unused intermediary accounts in `EndBlocker` of the superfluid module.

## Testing and Verifying

Added test cases
## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? yes
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? yes
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/superfluid/abci.go
Comment (mattverse) [inline x/superfluid/abci.go:23]: Question to reviewers: How do we set orderings of Endblockers between modules?

File: app/modules.go
Comment (p0mvn) [inline app/modules.go:169]:
```suggestion
	// have to be deleted before superfluid EndBlocker.
```

File: x/superfluid/module.go
Comment (p0mvn) [inline x/superfluid/module.go:190]: What is the reason for having this `EndBlocker` function? Why can't we run that logic directly here?

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:140]: nit: some of the points were read as "check by checking"
```suggestion
// We determine this by
// - using staking keeper to check if actual delegations exist,
// - checking if there are any intermediary account connection remaining to the intermediaryAcc
// - checking if there's no synthetic lock with the intermediaryAccount.Denom + intermediaryAccount.ValAddr combination
```

Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:140]: Another nit: I don't think that the method spec should be concerned about such implementation details. The API users only need to know what the method does / returns given the inputs.

The caller should not need to know such implementation details, especially when there are comments directly in code that say the same thing.

Please consider removing method spec implementation details while keeping the comments in code.

I'm not feeling strongly about this so feel free to resolve it if you disagree

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:147]:
```suggestion
	iterator := intermediaryAccConnectionPrefixStore.Iterator(nil, nil)
	defer iterator.Close()
```

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:159]: It feels to me that we can immediately return if we see an intermediate account existence.

```suggestion
	for ; iterator.Valid(); iterator.Next() {
		if sdk.AccAddress(iterator.Value()).Equals(intermedairyAcc.GetAccAddress()) {
			return true
		}
	}
```

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:162]: nit: noticing multiple calls to `intermedairyAcc.GetAccAddress()`. Does it make sense to have a variable for it?

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:171]: I think err check might be missed here

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:177]: ditto for error check

File: x/superfluid/keeper/stake_test.go
Comment (p0mvn) [inline x/superfluid/keeper/stake_test.go:539]: 0 by default
```suggestion
```

File: x/superfluid/keeper/stake_test.go
Comment (p0mvn) [inline x/superfluid/keeper/stake_test.go:612]: nit: `suite.Ctx` -> `ctx := suite.Ctx` and then reuse `ctx`

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:141]: Do we have a test specifically for this method?

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:113]: Do we have a direct test for this? I think we are only testing it vicariously via `EndBlocker`

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:141]: Does it make sense having this method exported?

File: x/superfluid/keeper/intermediary_account.go
Comment (p0mvn) [inline x/superfluid/keeper/intermediary_account.go:123]: Does this method need to be exported?

**Review by p0mvn — COMMENTED** [review]
Matt's explanation + implementation makes sense. Nice work on this!

I would like to request testing some methods directly. Other than that, most of my comments are nits.

File: x/superfluid/module.go
Comment (mattverse) [reply x/superfluid/module.go:190]: I don't know the exact reason behind this, but afaik, this is one of the conventions in settings endblockers, I'm guessing it's because we want to separate out logic for abci? 
cc @alexanderbez

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:123]: changed this to private!

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:140]: agreed with you that method spec does not need these information, but still thought having organized list of checks would be a useful comment to have. Moved these comments inside methods!

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:141]: This method is not changing state in any way, just checks if the given intermediary account has a check or not, I think it'd be nice to have this public as this method can come in handy in different places. lmk what you think

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:159]: Ah good catch on this!

File: x/superfluid/keeper/stake_test.go
Comment (mattverse) [reply x/superfluid/keeper/stake_test.go:539]: deciding to leave this as having this visible is better for test readability imo

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:171]: This is intentional, `GetSyntheticLockup` returns an error when it doesnt have the given lock, but in our case, this is not an error

We can also add a comment for this if this happens to be confusing.

I agree with you that this is bad API usage, an idea I would like to move forward to  after getting this PR merged is adding a method `HasSyntheticLockup` in the lockup keeper that returns a boolean for usage like this(we have other instances that uses `GetSyntheticLockup` like this whilst ignoring error, just to check if we have a synthetic lock)

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:177]: ditto for the reply above

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:113]: imo, we don't need direct test as the only thing `EndBlocker` is doing is calling `DeleteAllEmptyIntermediaryAccounts`. Even if we do have a test case, the test cases would look exactly the same.

I think the current test of `TestDeleteAllEmptyIntermediaryAccounts` should not be seen as a vicorous way of testing, but fully testing the original method + testing endblocker execution

lmk what you think

File: app/modules.go
Comment (alexanderbez) [inline app/modules.go:169]:
```suggestion
	// Superfluid should come after lockup module because matured synth locks
	// have to be deleted before superfluid EndBlock.
```

File: x/superfluid/keeper/epoch.go
Comment (alexanderbez) [inline x/superfluid/keeper/epoch.go:66]: just for context, why was this removed?

File: x/superfluid/module.go
Comment (alexanderbez) [reply x/superfluid/module.go:190]: It would be ugly to have that defined here IMO. Having it encapsulated in a function makes it easier to test.

Actually, now that I think of it, this really should be defined on the keeper itself, just like `InitGenesis`. I.e. `(k Keeper) EndBlock(...)`

File: x/superfluid/keeper/epoch.go
Comment (mattverse) [reply x/superfluid/keeper/epoch.go:66]: This line wasn't here in the first place, this was recently added to better log errors. However, with the main problem that caused all these errors fixed(by deleting intermediary account), we could revert it back to the code where we didn't have this line. 

Could still leave logging errors if necessary

File: x/superfluid/keeper/intermediary_account.go
Comment (mattverse) [reply x/superfluid/keeper/intermediary_account.go:137]: Added test cases!

**Review by alexanderbez — APPROVED** [review]

**Comment by alexanderbez** [comment]
I definitely feel like we should have a changelog entry for this.

**Comment by mattverse** [comment]
Current status of PR is that we're merging this upon checking logs for the current logs we have every epoch to confirm this PR would fix the main cause for the error logs

**Comment by mattverse** [comment]
Confirmed that the logs are actually happening because of the undeleted intermediary accounts.
Planning to merge this after getting the changelog entry and fixing the merge conflicts.

I don't think we should pressure getting it on chain asap imo as long as we have merged it.

**Comment by mattverse** [comment]
TODO For this PR pre-merge: write down edge casese that we think needs to be ensured testing, ensure that those cases are well unit tested

**Pull request closed without merging**
